File: Нейронные сети/lpc_nn/source/lpc55/irc2pc.f
******************************************************************
*
*	IRC2PC Version 48
*
******************************************************************
*
*   Convert Reflection Coefficients to Predictor Coeficients
*
* Inputs:
*  RC     - Reflection coefficients
*  ORDER  - Number of RC's
*  GPRIME - Excitation modification gain
* Outputs:
*  PC     - Predictor coefficients
*  G2PASS - Excitation modification sharpening factor
*
	SUBROUTINE IRC2PC( RC, PC, ORDER, GPRIME, G2PASS )
	INCLUDE 'config.fh'
	INTEGER ORDER, I, J
	REAL RC(ORDER), PC(ORDER), GPRIME, G2PASS, TEMP(MAXORD)

	G2PASS = 1.
	DO I = 1,ORDER
	   G2PASS = G2PASS*( 1. - RC(I)*RC(I) )
	END DO
	G2PASS = GPRIME*SQRT(G2PASS)
	PC(1) = RC(1)
	DO I = 2,ORDER
	   DO J = 1,I-1
	      TEMP(J) = PC(J) - RC(I)*PC(I-J)
	   END DO
	   DO J = 1,I-1
	      PC(J) = TEMP(J)
	   END DO
	   PC(I) = RC(I)
	END DO
	RETURN
	END

File: Нейронные сети/lpc_nn/source/lpc55-C/irc2pc.c
/******************************************************************
*
*	IRC2PC Version 48
*
******************************************************************
*
*   Convert Reflection Coefficients to Predictor Coeficients
*
* Inputs:
*  RC     - Reflection coefficients
*  ORDER  - Number of RC's
*  GPRIME - Excitation modification gain
* Outputs:
*  PC     - Predictor coefficients
*  *G2PASS - Excitation modification sharpening factor
*/

#include "config.ch"
#include "lpcdefs.h"
#include <math.h>

irc2pc( rc, pc, gprime, g2pass, where )
float rc[MAXORD][11], pc[], gprime, *g2pass;
int where;
{
int i,j;
float temp[MAXORD];

*g2pass = 1.;

for(i=0;i<ORDER;i++)	
	*g2pass = *g2pass*( 1. - rc[i][where]*rc[i][where] );

*g2pass = gprime*sqrt(*g2pass);
pc[0] = rc[0][where];

for(i=1;i<ORDER;i++)	{
	for(j=0;j<i;j++)
		temp[j] = pc[j] - rc[i][where]*pc[i-j-1];
	
	for(j=0;j<i;j++)
		pc[j] = temp[j];
	
	pc[i] = rc[i][where];
}

}
